File: solutions/cpp/0138.cpp
class Solution {
 public:
  Node* copyRandomList(Node* head) {
    if (!head) return nullptr;
    if (map.count(head)) return map[head];

    map[head] = new Node(head->val);
    map[head]->next = copyRandomList(head->next);
    map[head]->random = copyRandomList(head->random);

    return map[head];
  }

 private:
  unordered_map<Node*, Node*> map;
};

File: solutions/java/0138.java
class Solution {
  public Node copyRandomList(Node head) {
    if (head == null)
      return null;
    if (map.containsKey(head))
      return map.get(head);

    map.put(head, new Node(head.val));
    map.get(head).next = copyRandomList(head.next);
    map.get(head).random = copyRandomList(head.random);

    return map.get(head);
  }

  private Map<Node, Node> map = new HashMap<>();
}